drivers/counter/counter_gd32_timer.c

```diff
@@ -0,0 +1,534 @@
+	} else if (read(dev) >= cfg->ticks) {
+		err = -ETIME;
+		if (cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
+			set_counter(dev, 0);
+		}
+	}
+
+	if (cfg->callback) {
+		interrupt_enable(dev, TIMER_INT_UP);
+	}
+
+	return err;
+}
+
+static uint32_t get_pending_int(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+
+	return cfg->get_irq_pending();
+}
+
+static uint32_t get_freq(const struct device *dev)
+{
+	struct counter_gd32_data *data = dev->data;
+
+	return data->freq;
+}
+
+static uint32_t get_guard_period(const struct device *dev, uint32_t flags)
+{
+	struct counter_gd32_data *data = dev->data;
+
+	return data->guard_period;
+}
+
+static int set_guard_period(const struct device *dev, uint32_t guard,
+			    uint32_t flags)
+{
+	struct counter_gd32_data *data = dev->data;
+
+	__ASSERT_NO_MSG(guard < get_top_value(dev));
+
+	data->guard_period = guard;
+	return 0;
+}
+
+static void top_irq_handle(const struct device *dev)
+{
+	struct counter_gd32_data *data = dev->data;
+	counter_top_callback_t cb = data->top_cb;
+
+	if (interrupt_flag_get(dev, TIMER_INT_FLAG_UP) != 0) {
+		interrupt_flag_clear(dev, TIMER_INT_FLAG_UP);
+		__ASSERT(cb != NULL, "top event enabled - expecting callback");
+		cb(dev, data->top_user_data);
+	}
+}
+
+static void alarm_irq_handle(const struct device *dev, uint32_t chan)
+{
+	struct counter_gd32_data *data = dev->data;
+	struct counter_gd32_ch_data *alarm = &data->alarm[chan];
+	counter_alarm_callback_t cb;
+	bool hw_irq_pending = !!(interrupt_flag_get(dev, TIMER_FLAG_CH(chan)));
+	bool sw_irq_pending = data->cc_int_pending & TIMER_INT_CH(chan);
+
+	if (hw_irq_pending || sw_irq_pending) {
+		atomic_and(&data->cc_int_pending, ~TIMER_INT_CH(chan));
+		interrupt_disable(dev, TIMER_INT_CH(chan));
+		interrupt_flag_clear(dev, TIMER_FLAG_CH(chan));
+
+		cb = alarm->callback;
+		alarm->callback = NULL;
+
+		if (cb) {
+			cb(dev, chan, read(dev), alarm->user_data);
+		}
+	}
+}
+
+static void irq_handler(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+
+	top_irq_handle(dev);
+
+	for (uint32_t i = 0; i < cfg->counter_info.channels; i++) {
+		alarm_irq_handle(dev, i);
+	}
+}
+
+static int counter_gd32_timer_init(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+	struct counter_gd32_data *data = dev->data;
+	uint32_t pclk;
+
+	clock_control_on(GD32_CLOCK_CONTROLLER,
+			 (clock_control_subsys_t *)&cfg->clkid);
+	clock_control_get_rate(GD32_CLOCK_CONTROLLER,
+			       (clock_control_subsys_t *)&cfg->clkid, &pclk);
+
+	data->freq = pclk / (cfg->prescaler + 1);
+
+	interrupt_disable(dev, TIMER_INT_ALL);
+	reset_line_toggle_dt(&cfg->reset);
+
+	cfg->irq_config(dev);
+	set_prescaler(dev, cfg->prescaler);
+	set_autoreload_value(dev, cfg->counter_info.max_top_value);
+	set_software_event_gen(dev, TIMER_SWEVG_UPG);
+
+	return 0;
+}
+
+static const struct counter_driver_api counter_api = {
+	.start = start,
```

Comment:
Please add name prefix `counter_gd32_` to these api implements.

Reply:
May futurely add the RTC driver.
Add `counter_gd32_timer_` as a prefix.